Mathlib/Data/Matroid/Dual.lean

```diff
@@ -0,0 +1,260 @@
+    · rw [and_iff_right hI'X]
+      refine' disjoint_of_subset_right hB'IB _
+      rw [disjoint_union_right, and_iff_left hI'B]
+      exact disjoint_of_subset hI'X hI.subset disjoint_sdiff_right
+    simp only [mem_setOf_eq, subset_inter_iff, and_imp, forall_exists_index]
+    intros J hJE B'' hB'' hdj _ hJX hssJ
+    rw [and_iff_left hJE]
+    rw [diff_eq, inter_right_comm, ←diff_eq, diff_subset_iff] at hssJ
+
+    have hI' : (B'' ∩ X) ∪ (B' \ X) ⊆ B'
+    · rw [union_subset_iff, and_iff_left (diff_subset _ _),
+        ← inter_eq_self_of_subset_left hB''.subset_ground, inter_right_comm, inter_assoc]
+
+      calc _ ⊆ _ := inter_subset_inter_right _ hssJ
+           _ ⊆ _ := by rw [inter_distrib_left, hdj.symm.inter_eq, union_empty]
+           _ ⊆ _ := inter_subset_right _ _
+
+    obtain ⟨B₁,hB₁,hI'B₁,hB₁I⟩ := (hB'.indep.subset hI').exists_base_subset_union_base hB''
+    rw [union_comm, ←union_assoc, union_eq_self_of_subset_right (inter_subset_left _ _)] at hB₁I
+
+    have : B₁ = B'
+    · refine hB₁.eq_of_subset_indep hB'.indep (fun e he ↦ ?_)
+      refine (hB₁I he).elim (fun heB'' ↦ ?_) (fun h ↦ h.1)
+      refine (em (e ∈ X)).elim (fun heX ↦ hI' (Or.inl ⟨heB'', heX⟩)) (fun heX ↦ hIB' ?_)
+      refine hI.mem_of_insert_indep ⟨hB₁.subset_ground he, heX⟩
+        (hB₁.indep.subset (insert_subset he ?_))
+      refine (subset_union_of_subset_right (subset_diff.mpr ⟨hIB',?_⟩) _).trans hI'B₁
+      refine disjoint_of_subset_left hI.subset disjoint_sdiff_left
+
+    subst this
+
+    refine' subset_diff.mpr ⟨hJX, by_contra (fun hne ↦ _)⟩
+    obtain ⟨e, heJ, heB'⟩ := not_disjoint_iff.mp hne
+    obtain (heB'' | ⟨-,heX⟩ ) := hB₁I heB'
+    · exact hdj.ne_of_mem heJ heB'' rfl
+    exact heX (hJX heJ)
+  subset_ground := by tauto
+
+/-- The dual of a matroid; the bases are the complements (w.r.t `M.E`) of the bases of `M`. -/
+def dual (M : Matroid α) : Matroid α := M.dualIndepMatroid.matroid
+
+/-- The `﹡` symbol, which denotes matroid duality.
+  (This is distinct from the usual `*` symbol for multiplication, due to precedence issues. )-/
+postfix:max "﹡" => Matroid.dual
+
+theorem dual_indep_iff_exists' : (M﹡.Indep I) ↔ I ⊆ M.E ∧ (∃ B, M.Base B ∧ Disjoint I B) := by
+  simp [dual]
+
+@[simp] theorem dual_ground : M﹡.E = M.E := rfl
+
+@[simp] theorem dual_indep_iff_exists (hI : I ⊆ M.E := by aesop_mat) :
+    M﹡.Indep I ↔ (∃ B, M.Base B ∧ Disjoint I B) := by
+  rw [dual_indep_iff_exists', and_iff_right hI]
+
+theorem dual_dep_iff_forall : (M﹡.Dep I) ↔ (∀ B, M.Base B → (I ∩ B).Nonempty) ∧ I ⊆ M.E := by
+  simp_rw [dep_iff, dual_indep_iff_exists', dual_ground, and_congr_left_iff, not_and,
+    not_exists, not_and, not_disjoint_iff_nonempty_inter, imp_iff_right_iff, iff_true_intro Or.inl]
+
+instance dual_finite [M.Finite] : M﹡.Finite :=
+  ⟨M.ground_finite⟩
+
+instance dual_nonempty [M.Nonempty] : M﹡.Nonempty :=
+  ⟨M.ground_nonempty⟩
+
+@[simp] theorem dual_base_iff (hB : B ⊆ M.E := by aesop_mat) : M﹡.Base B ↔ M.Base (M.E \ B) := by
```

Comment:
Naming convention warning: this would often be `base_dual_iff`, since at least for data, the fact you used dot notation isn't taken into account for the order.

This sometimes goes both ways, and also `@[simps]` doesn't generate the "right" order, so I'll let it go for now, but someone might come along and rename it in a future refactoring.

Reply:
ok, thanks!